text/0072-kafka-schema-registry.md

```diff
@@ -0,0 +1,113 @@
+- Start Date: 2023-02-01
+- RFC Type: feature
+- RFC PR: n/a
+- RFC Status: draft
+
+# Summary
+
+This RFC proposes introducing a centralized schema repository for the Kafka topics in use at Sentry. It provides mappings between the Kafka topics we are using to the encoding format and schemas for messages on that topic.
+
+Topic registration and schema validation of messages will be optional for the foreseeable future.
+
+# Motivation
+
+Kafka is increasingly used as the message bus between services at Sentry. Kafka topics and their schemas are usually not internal to any one service, but part of the contract between services.
+
+As Sentry grows, the number of message types and topics in use is exploding. We are marching further towards increasingly distributed ownership of data by product and infrastructure engineering teams. We have already seen an increase in the number of incidents related with invalid data and schema issues at Sentry. It is reasonable to expect this trend to continue as the number of topics, data types, teams and engineers contributing to Sentry grows.
+
+The goals of the centralized schema repository are to:
+
+- provide an explicit, single source of truth about how data in any given topic should look, and the contract that all consumers and producers of a topic must adhere to
+- provide more stability by enforcing backward compatibility of changes via automation
+- explicitly enumerate ownership of schemas
+- make it easier for consumers to identify and reject bad messages without pausing the whole pipeline
+- provide examples of messages for each schema type, which would help with writing consumer tests
+
+# Supporting Data
+
+Over the last quarter, we have seen many incidents related to schema disagreements at Sentry. Examples include the following: (note: inc numbers are internal to Sentry)
+
+- Post process forwarder (INC-218)
+- Snuba’s transactions consumer (INC-210)
+- Super big consumers (INC-220)
+- Replays consumer (INC-250, INC-281)
+
+Many of these incidents are P1. Since messages are in order, an invalid message often halts consumers and requires manual intervention. This is disruptive to both Sentry engineers and our users (it takes us much longer to recover).
+
+# Options Considered
+
+### **Option A (preferred): Publish a library**
+
+Schemas will be made available as a library for multiple programming languages (Python and Rust to start).
```

Comment:
A concern was raised during the TSC discussion about the inadequacy of the tooling we have today to handle cross repo dependencies. (think about the pain when cutting new arroyo releases).

I think we need to mitigate that issue and, specifically for this RFC we should outline the process that we need to follow when applying schema changes in order to propagate them to all clients of the library.

I would try to answer these questions:
- versioning and releasing. Are we going to follow the same releasing process we have for other internal libraries like Arroyo? How do we make it seamless?
- How do we encourage clients to stay up to date? Are we going to build automation in all consumers? Or are we going to make arroyo provide tools to fetch the recent schema?
- Is there a revert process in case we sneak a bad change into the repo by mistake?


I would focus on keeping the process as light and as seamless as possible. If there is too much bureaucracy to begin with, there will be resistance in adopting the schema itself.

Reply:
> * versioning and releasing. Are we going to follow the same releasing process we have for other internal libraries like Arroyo? How do we make it seamless?

The plan is to use Sentry standard releasing processes. Yes, there is some friction with approvers, etc that already exists today. But given these problems already exist today, I think we need to focus on solving this for all Sentry libraries not build anything special here.

> * How do we encourage clients to stay up to date? Are we going to build automation in all consumers? Or are we going to make arroyo provide tools to fetch the recent schema?

As mentioned above, I think in the short term we can just add some aggressive dependabot config on all the services that use the library.

> * Is there a revert process in case we sneak a bad change into the repo by mistake?

I don't think we should have any special revert process. Since this is a library and not a service, we don't need the kind of "fast revert" that some other Sentry repositories have. I worry that it would complicate the rules around compatibility and versioning, so I'd vote to just roll forward to undo the change.

Reply:
> But given these problems already exist today, I think we need to focus on solving this for all Sentry libraries not build anything special here.

I don't think I agree that this is a problem that exists today. We do not have a schema registry today, changing the schema does not require any external approval just to perform the release. Rely on the current process can introduce this additional step that does not add any value to the quality of the schema. This will be unneeded friction. 

Could you please explore whether we can allow every owner of a schema to be allowed to cut releases ? I think that would be an ok process.
Imagine if the replays team adds a field to their schema. If they can cut a release without external approval we would be ok with the current tool. If they have to wait for somebody potentially in a different time zone that is a non starter.

Reply:
Confirmed with devinfra that internal packages will get autoapproval and not require the sam release process as SDKs and customer installed packages

Reply:
Mind writing your last statement in the RFC as well. Let's make it clear that we target a lighter release process than what we do today.